src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs

```diff
@@ -97,7 +99,22 @@ protected override bool OnCollect(int timeoutMilliseconds)
         /// <inheritdoc />
         protected override bool OnShutdown(int timeoutMilliseconds)
         {
-            return this.exporter.Shutdown(timeoutMilliseconds);
+            var result = true;
+
+            if (timeoutMilliseconds == Timeout.Infinite)
+            {
+                result = this.Collect(Timeout.Infinite) && result;
+                result = this.exporter.Shutdown(Timeout.Infinite) && result;
+            }
+            else
+            {
+                var sw = Stopwatch.StartNew();
+                result = this.Collect(timeoutMilliseconds) && result;
+                var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;
+                result = this.exporter.Shutdown((int)Math.Max(timeout, 0)) && result;
```

Comment:
we'll need to modify PrometheusExporter to override shutdown, and shut its httpserver, right? 
And obviously, for pull based exporters, any pending data is simply lost on shutdown right?

Reply:
Yes and yes.